src/acquisition/covidcast_nowcast/load_sensors.py

```diff
@@ -33,17 +36,20 @@ def main(csv_path: str = SENSOR_CSV_PATH) -> None:
     """
     user, pw = secrets.db.epi
     engine = sqlalchemy.create_engine(f"mysql+pymysql://{user}:{pw}@{secrets.db.host}/{DB_NAME}")
-    for filepath, attributes in CsvImporter.find_csv_files(csv_path):
-        if attributes is None:
-            move(filepath, filepath.replace("receiving", "archive/failed"))
+    for filepath, attr in CsvImporter.find_csv_files(csv_path):
+        if attr is None:
+            _move_after_processing(filepath, success=False)
             continue
         try:
-            data = load_and_prepare_file(filepath, attributes)
-            data.to_sql(TABLE_NAME, engine, if_exists="append", index=False)
+            data = load_and_prepare_file(filepath, attr)
+            conn = engine.connect()
+            with conn.begin():
+                method = _create_upsert_method(sqlalchemy.MetaData(conn))
+                data.to_sql(TABLE_NAME, engine, if_exists="append", method=method, index=False)
         except Exception:
-            move(filepath, filepath.replace("receiving", "archive/failed"))
+            _move_after_processing(filepath, success=False)
             raise
```

Comment:
Do you need a `finally` block to do any connection cleanup with the database? I know the mysql.connect library requires something like
https://github.com/cmu-delphi/delphi-epidata/blob/e6715c0602d889324f48153f6ad5e02409d678d8/src/acquisition/covidcast/csv_to_database.py#L179

but not sure if sqlalchemy needs something similar

Reply:
i wondered that too when i first implemented it, since ive had to cloes sqlalchemy connections before, but the to_sql method can take an engine and not a connection. Based on [this answer](https://stackoverflow.com/a/51242577), looks like it isn't needed. The connection used for the connection metadata is in a context manager so i think thats ok as well